waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,379 @@
+                                 "Must have exactly one assertion but got " (count assertions))
+                            {:status 400})))
+        assertion (first assertions)
+        _ (when-not (saml-assertion-signature-valid? assertion idp-cert)
+            (throw (ex-info "Could not authenticate user. Invalid SAML assertion signature."
+                            {:status 400})))
+        {:keys [attrs confirmation name-id]} (parse-saml-assertion assertion)
+        not-on-or-after (clj-time.coerce/from-sql-time (:not-on-or-after confirmation))
+        t-now (t/now)
+        _ (when-not (t/before? t-now not-on-or-after)
+            (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                            {:status 400
+                             :expiry-time not-on-or-after
+                             :current-time t-now})))
+        email (first (get attrs "email"))
+        ; https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/technical-reference/the-role-of-claims
+        upn (first (get attrs "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn"))
+        name-id-value (get name-id :value)
+        saml-principal (or upn name-id-value)
+        {:keys [authorization/principal authorization/user]} (auth/auth-params-map saml-principal)
+        saml-principal' (if (and email (= principal user)) email saml-principal)
+        saml-auth-data (utils/map->base-64-string
+                         {:not-on-or-after not-on-or-after :redirect-url request-url :saml-principal saml-principal'}
+                         password)]
+    {:body (render-authenticated-redirect-template {:auth-redirect-uri (str scheme "://" host auth-redirect-endpoint)
+                                                    :saml-auth-data saml-auth-data})
+     :status 200}))
+
+(defn saml-auth-redirect-handler
+  "Endpoint for POSTs to Waiter with IdP-signed credentials. If signature is valid, return principal and original request."
+  [request {:keys [password]}]
+  {:pre [(not-empty password)]}
+  (if-let [saml-auth-data (get-in (ring-params/params-request request) [:form-params "saml-auth-data"])]
+    (let [{:keys [not-on-or-after redirect-url saml-principal]}
+          (try
+            (utils/base-64-string->map saml-auth-data password)
+            (catch Exception e
+              (throw (ex-info "Could not parse saml-auth-data." {:status 400
+                                                                 :saml-auth-data saml-auth-data
+                                                                 :inner-exception e} e))))
+          t-now (t/now)
+          _ (when-not (t/before? t-now not-on-or-after)
+              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                              {:status 400
+                               :saml-assertion-not-on-or-after not-on-or-after
+                               :t-now t-now})))
```

Comment:
Can we use `expiry-time` and `current-time` as keys?

Reply:
done